ANTLR/src/Compiler/ErrorHandling/UnderlineErrorListener.java

```diff
@@ -0,0 +1,166 @@
+     * @param errorMsg Message that explains the cause of the error
+     * @param offendingToken A tokens which will be underlined
+     * @param additionalOffendingTokens The additional tokens which will be underlined
+     */
+    public void ThrowUnderlinedError(String errorMsg, Token offendingToken, Token... additionalOffendingTokens) {
+        ArrayList<Token> allOffendingTokens = new ArrayList<Token>();
+        allOffendingTokens.add(offendingToken);
+
+        for (Token token : additionalOffendingTokens){
+            allOffendingTokens.add(token);
+        }
+
+        ThrowUnderlinedError(errorMsg, allOffendingTokens);
+    }
+
+    /**
+     * Converts a list of tokens to a list of LineWithError's. Checks whether or not offending tokens are on the same line.
+     * @param offendingTokens A tokens which will be underlined
+     * @return An ArrayList of LineWithError.
+     */
+    private ArrayList<LineWithError> GetLineWithErrorArray(ArrayList<Token> offendingTokens) {
+        ArrayList<LineWithError> errorLines = new ArrayList<LineWithError>();
+
+        for (Token offendingToken : offendingTokens) {
+            int lineNumber = offendingToken.getLine();
+
+            //Checks whether or not an existing offending token is present in a line.
+            LineWithError errorLine = null;
+            for (LineWithError line : errorLines){
+                if (line.lineNumber == lineNumber){
+                    errorLine = line;
+                }
+            }
+
+            if (errorLine == null){
+                //Gets the source code for the line in which the offending token occurred
+                String line = offendingToken.getInputStream().toString().split("\n")[lineNumber - 1];
+                LineWithError lineWithError = new LineWithError(line, lineNumber);
+                lineWithError.addTokenToInterval(offendingToken);
+                errorLines.add(lineWithError);
+            }
+            else {
+                errorLine.addTokenToInterval(offendingToken);
+            }
+        }
+        return errorLines;
+    }
+
+    /**
+     * Prints a line of source code, the line number and hightligts the offending tokens
+     * @param lineWithError The line that should be printed
+     */
+    private void PrintUnderlinedError(LineWithError lineWithError) {
+        System.out.println("At line " + lineWithError.lineNumber + ":\n");
+        System.out.println(lineWithError.line);
+        printUnderlinedIntervals(lineWithError.errorIntervals);
+    }
+
+    /**
+     * Prints out the "^" symbol under offending tokens
+     * @param offendingTokenInterval
+     */
+    private void printUnderlinedIntervals(ArrayList<Interval> offendingTokenInterval) {
+        int printPos = 0;
+
+        for (Interval interval : offendingTokenInterval){
+            while (printPos < interval.a) {
+                System.out.print(" ");
+                printPos++;
+            }
```

Comment:
I would like to see this abstracted away into an auxiliary method. Also, I'd like to see print only being called once.

Reply:
Or maybe even better - just don't loop at all.
```suggestion
            if (printPos < interval.a) {
                System.out.print(" ".repeat(interval.a - printPos));
                printPos = interval.a;
            }
```

Reply:
ok